rust/src/producer.rs

```diff
@@ -0,0 +1,171 @@
+//! This module provides an abstraction over a Kafka producer in
+//! order to allow client code to instantiate the producer
+//! implementation they want without depending on the rdkafka
+//! ThreadedProducer.
+//!
+//! It also simplify unit tests.
+
+use rdkafka::config::ClientConfig as RdKafkaConfig;
+use rdkafka::producer::{BaseRecord, ThreadedProducer};
+use rdkafka::producer::{DeliveryResult, ProducerContext};
+use rdkafka::ClientContext;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
+use thiserror::Error;
+
+/// This structure wraps the parameters to initialize a producer.
+/// This struct is there in order not to expose the rdkafka
+/// details outside.
+#[derive(Debug, Clone)]
+pub struct KafkaConfig {
+    config_map: HashMap<String, String>,
+}
+
+impl KafkaConfig {
+    pub fn new_producer_config(
+        bootstrap_servers: &str,
+        override_params: Option<HashMap<String, String>>,
+    ) -> Self {
+        let mut config_map: HashMap<String, String> = HashMap::new();
+        config_map.insert("bootstrap.servers".to_string(), bootstrap_servers.into());
+
+        let config = Self { config_map };
+
+        apply_override_params(config, override_params)
+    }
+}
+
+impl From<KafkaConfig> for RdKafkaConfig {
+    fn from(item: KafkaConfig) -> Self {
+        let mut config_obj = RdKafkaConfig::new();
+        for (key, val) in item.config_map.iter() {
+            config_obj.set(key, val);
+        }
+        config_obj
+    }
+}
+
+fn apply_override_params<V>(
+    mut config: KafkaConfig,
+    override_params: Option<HashMap<String, V>>,
+) -> KafkaConfig
+where
+    V: Into<String>,
+{
+    if let Some(params) = override_params {
+        for (param, value) in params {
+            config.config_map.insert(param, value.into());
+        }
+    }
+    config
+}
+
+pub struct CaptureErrorContext;
+
+impl ClientContext for CaptureErrorContext {}
+
+impl ProducerContext for CaptureErrorContext {
+    type DeliveryOpaque = ();
+
+    fn delivery(&self, _: &DeliveryResult, _delivery_opaque: Self::DeliveryOpaque) {
+        // TODO: Do something useful. AS of now guarantees of delivery
+        // in this producer are not particularly critical.
+    }
+}
+
+/// Kafka producer errors.
+#[derive(Error, Debug)]
+pub enum ClientError {
+    /// Failed to send a kafka message.
+    #[error("failed to send kafka message")]
+    SendFailed(#[source] rdkafka::error::KafkaError),
+
+    /// Failed to create a kafka producer because of the invalid configuration.
+    #[error("failed to create kafka producer: invalid kafka config")]
+    InvalidConfig(#[source] rdkafka::error::KafkaError),
+}
+
+/// A basic Kafka Producer trait.
+///
+/// We do not neet to set headers or key for this data.
+pub trait Producer {
+    fn send(&mut self, topic_name: &str, payload: &[u8]) -> Result<(), ClientError>;
+}
+
+pub struct DummyProducer {
```

Comment:
I'd move this into the test

Reply:
Cannot move it into the test. 
I will make it `#[cfg(test)]`.  and `pub(crate)`
This is needed by the test of another module in the crate (accountant.rs)